Lesson29/Gokhan/script.js

```diff
@@ -0,0 +1,163 @@
+console.log("Ex4: ", fullName1);
+
+
+
+// 5. Nested Object Destructuring
+// Given the following nested object, use object destructuring to assign the values of name, age, and city to their respective variables:
+const personE17 = {
+	name: 'John Doe',
+	age: 30,
+	address: {
+		city: 'New York',
+	},
+};
+
+const { name, age: age1, address } = personE17;
+const { city } = address;
+console.log("Ex5: ", ' name: ', name, ' age: ', age1, ' city: ', city);
+
+
+
+// 6. Default Parameters + Arrow function
+// Convert Named Function to Arrow Function with Default Parameters
+// function greet(name, greeting) {
+// 	return `${greeting}, ${name}!`;
+// }
+const greet = (greeting='hello', name='Aysel') => `${greeting}, ${name}!`;
+console.log("Ex6: ",greet('hi'));
+
+
+
+// 7. Default Parameters
+// Add Default Parameters to an Existing Arrow Function, Default tax rate 0.1, default discount is 0.
+const calculateTotal = (price, taxRate = 0.1, discount = 0) => {
+	return price + price * taxRate - discount;
+};
+
+const calculateTotalVAT10 = (price, discount = 0) => {
+  return calculateTotal(price, 0.1, discount);
+}
+
+console.log("Ex7: ",calculateTotalVAT10(100, 0));
+
+
+
+// Optional Chaining
+// Optional chaining allows you to safely access deeply nested properties.
+
+// 8. Safe Access to Nested Object Properties
+// Update the code to safely access userName and userCity using optional chaining to handle cases where properties might be missing.
+const userEx8 = {
+	profile: {// optional
+		name: 'Alice',
+		address: {
+			city: 'Wonderland',
+		},
+	},
+};
+const userName = userEx8?.profile?.name;
+console.log('EX8: ', userName);
+const userCity = userEx8?.profile?.address?.city;
+console.log('EX8: ', userCity);
+
+
+
+// 9. Handle Missing Properties
+// Update the code to use optional chaining to safely access userCountry and provide a default value of 'Unknown' if the property is missing.
+
+const userEx9 = {
+	profile: {
+		name: 'Alice',
+    address: {
+      country: null
+    }
+	},
+};
+
+const userCountry = userEx9?.profile?.address?.country || 'Unknown'; // left (if true) OR right
+console.log("Ex9: ",userCountry);
+
+
+
+// 10. Optional Chaining with Function Calls
+// Update the code to safely call the getName function using optional chaining, considering that profile or getName might be missing.
+
+const userEx10 = {
+	profile: {
+		getName: () => 'Alice',
+	},
+};
+
+const safelyCall =userEx10?.profile?.getName();
+console.log('Ex10: ', safelyCall);
+
+
+
+// 11. Rewrite the code using the nullish coalescing operator to assign a default value to storedData only if userInput is null or undefined.
+let userInput = 0;
+let storedData = userInput ?? 'Default Value';
+
+
+
+console.log( "Ex11: ", "storedData: " , storedData); // Default Value
+
+
+
+// 12. Rewrite the code using the nullish coalescing operator to display number of users even if it is 0.
+let userCount = 0;
+let displayCount = userCount || 'No users';
```

Comment:
```suggestion
let displayCount = userCount ?? 'No users';
```

Nullish coalescing operator (??) should be used instead.